ecs-init/gpu/nvidia_gpu_manager.go

```diff
@@ -0,0 +1,218 @@
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package gpu
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/NVIDIA/gpu-monitoring-tools/bindings/go/nvml"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+//go:generate mockgen.sh $GOPACKAGE $GOFILE
+
+// GPUManager encompasses methods to get information on GPUs and their driver
+type GPUManager interface {
+	Setup() error
+	Initialize() error
+	Shutdown() error
+	GetGPUDeviceIDs() ([]string, error)
+	GetDriverVersion() (string, error)
+	DetectGPUDevices() error
+	SaveGPUState() error
+}
+
+// NvidiaGPUManager is used as a wrapper for NVML APIs and implements GPUManager
+// interface
+type NvidiaGPUManager struct {
+	DriverVersion       string
+	NvidiaDockerVersion string
+	GPUIds              []string
+}
+
+const (
+	// NvidiaGPUDeviceFilePattern is the pattern of GPU device files on the instance
+	NvidiaGPUDeviceFilePattern = "/dev/nvidia*"
+	// GPUInfoDirPath is the directory where gpus and driver info are saved
+	GPUInfoDirPath = "/var/lib/ecs/gpu"
+	// NvidiaGPUInfoFilePath is the file path where gpus and driver info are saved
+	NvidiaGPUInfoFilePath = "/var/lib/ecs/gpu/nvidia-gpu-info.json"
+	// FilePerm is the file permissions for gpu info json file
+	FilePerm = 0700
+)
+
+// ErrNoGPUDeviceFound is thrown when it is not a ECS GPU instance
+var ErrNoGPUDeviceFound = errors.New("No GPU device files found in the instance")
+
+// NewNvidiaGPUManager is used to obtain NvidiaGPUManager handle
+func NewNvidiaGPUManager() GPUManager {
+	return &NvidiaGPUManager{}
+}
+
+// Setup is used for setting up gpu information in the instance
+func (n *NvidiaGPUManager) Setup() error {
+	err := n.DetectGPUDevices()
+	if err != nil {
+		if err == ErrNoGPUDeviceFound {
+			return nil
+		}
+		return errors.Wrapf(err, "setup failed")
+	}
+	err = n.Initialize()
+	if err != nil {
+		return errors.Wrapf(err, "setup failed")
+	}
+	defer n.Shutdown()
+	version, err := n.GetDriverVersion()
+	if err != nil {
+		return errors.Wrapf(err, "setup failed")
+	}
+	n.DriverVersion = version
+	gpuIds, err := n.GetGPUDeviceIDs()
+	if err != nil {
+		return errors.Wrapf(err, "setup failed")
+	}
+	n.GPUIds = gpuIds
+	err = n.SaveGPUState()
+	if err != nil {
+		return errors.Wrapf(err, "nvidia gpu manager: setup failed")
+	}
+	return nil
+}
+
+// DetectGPUDevices checks if GPU devices are present in the instance
+func (n *NvidiaGPUManager) DetectGPUDevices() error {
+	matches, err := MatchFilePattern(NvidiaGPUDeviceFilePattern)
+	if err != nil {
+		return errors.Wrapf(err, "nvidia gpu manager: detecting GPU devices failed")
+	}
+	if matches == nil {
+		return ErrNoGPUDeviceFound
+	}
+	return nil
+}
+
+var MatchFilePattern = FilePatternMatch
```

Comment:
Why `MatchFilePattern` cannot be `matchFilePattern`? seems like we don't use it outside of this package.